### ISSUE TITLE
Question about the time complexity.

### ISSUE THREAD

**Issue by Tonyhukaifan** [sands-lab/grace#5]
@hangxu0304 I mixed two methods (TopK and QSGD) to make a compression on gradient vector. But I found that it needs much more time in one epoch during training than a single way to do that. I can't find the reason. The code about the compressor are shown below:
```
    def compress(self, tensor, name):

        shape = tensor.size()
        numel = tensor.numel()
        tensor = tensor.flatten()

        k = max(1, int(numel * self.compress_ratio))
        _, indices = torch.topk(tensor.abs(), k)

        values = tensor[indices]

        ################################################################################################################
        upperbound = torch.max(values.abs()).flatten()

        abs_gradient = values.abs()

        abs_gradient *= 127 / upperbound

        now_level = abs_gradient.floor()
        prob = torch.empty_like(values).uniform_()
        is_next_level = (prob < (abs_gradient - now_level)).type(torch.float32)
        now_level += is_next_level

        tensor_compressed = (now_level * values.sign()).type(torch.int16)
        tensor_compressed = tensor_compressed.type(torch.int8)

        ctx = numel, shape

        return [tensor_compressed, indices, upperbound], ctx

    def decompress(self, tensor_compressed, ctx):

        numel, shape = ctx
        values, indices , upperbound = tensor_compressed

        decode_output = values.type(torch.float32).abs()

        decode_output *= upperbound / 127
        decode_output *= values.sign()

        tensor_decompressed = torch.zeros(numel, dtype=decode_output.dtype, layout=decode_output.layout, device=decode_output.device)
        tensor_decompressed.scatter_(0, indices, decode_output)

        return tensor_decompressed.view(shape)
```

And I also make an experiment on the` compress` and `decompress` function to check if these have high computational complexity. But I found that they take less time to deal with a vector of which dimension up to 1,000,000. The time cost is just about 0.002 seconds.

Appreciate for your advice and help !!！
Sincerely yours.

**Comment by hangxu0304** [comment]
Hi,

I did some benchmarks in our own environment (2 nodes with one Nvidia v100 GPU, torch 1.3.1, ResNet20, Cifar10, batch_size=256). Here are the time of 1 epoch training for different compressors:
- NoneCompressor(allreduce): 11.03 s
- TopKCompressor(0.01)(allgather): 11.35 s
- QSGDCompressor(127)(allgather): 11.91 s
- YourCode(0.01)(allgather):   14.11 s

Looks like the time cost of your code is much higher than either of Topk and QSGD.
Then I add a simple loop in TopKCompressor and QSGDCompressor to simulate high computation overhead:

|         Number of loops         | 1     | 2     | 3     | 4     |
|:-------------------------------:|-------|-------|-------|-------|
| TopKCompressor(0.01)(allgather) | 11.35 | 12.80 | 13.25 | 13.60 |
|  QSGDCompressor(127)(allgather) | 11.91 | 15.32 | 18.87 | 22.46 |


The training time grows up linearly with computation overhead, and we can derive the computation cost for TopK and QSGD in this case,  which are 0.45 s and 3.4 s for 1 epoch training. Then let's look back at the time cost of your code, it makes more sense now. I think the reason why single TopK or QSGD doesn't show high time cost is probably the benefit of less communication. As you increase the computation overhead, the communication benefits become negligible.

**Comment by Tonyhukaifan** [comment]
Thanks for your reply.

I also did some experiments on the GRACE(change the flag `tensors_size_are_same` in Compressor Init.), I found that the time cost is much higher when we turn the `tensors_size_are_same` to `False`. So I think it will bring us higher computational complexity when dealing with different sizes of the gradient tensor.

**Comment by hangxu0304** [comment]
Indeed. When `tensors_size_are_same` is set to False, GRACE needs an additional allgather for the size of the tensors, that's why it is slow.
